File: raw2FullLFES.m
function myLFES = raw2FullLFES(myLFES)
%%  Copyright 2018 Engineering Systems Analytics LLC
%
% This function converts an "raw" data LFES into a "full" LFES.
% Proposed Usage:  myLFES=raw2FullLFES(myLFES)
disp(['I am entering ' 'raw2FullLFES.m'])
%% Calculate Resource Indices
myLFES=calcResourceIndices(myLFES);

%% Calc Resource Counts
myLFES=calcResourceCounts(myLFES);

%% Calc Resource set 
myLFES=calcResourceSet(myLFES);

%% Pack Set of Transformation Processes
if myLFES.DOFM>0
    myLFES=packSetTransformProcess(myLFES);
end

%% Make Set of Transportation Processes
if myLFES.DOFH>0
   myLFES=makeSetTransportProcess(myLFES);
end

%% Make Set of Refined Transportation Processes
if myLFES.DOFH>0
   myLFES=makeSetTransportRefProcess(myLFES);
end
%% Initialize Knowledge Bases and Constraint Matrices
myLFES = initializeKnowledgeBasesConstraintsMatrices(myLFES);

%% Calculate Transformation Knowledge Base JM, Constraints Matrix KM, and Transformation Concept AM
if myLFES.DOFM>0
    myLFES = calcJM_KM_AM(myLFES);
end

%% Calc Indices 
if myLFES.numMachines > 0
    myLFES = calcMachineIdxPort(myLFES);
end
if myLFES.numIndBuffers > 0
    myLFES = calcIndBufferIdxPort(myLFES);
end
if myLFES.numTransporters > 0
    myLFES = calcTransporterIdxPort(myLFES);
end

%% Calculate Transportation Knowledge Base JH, Constraints Matrix KH, and Transformation Concept AH
if myLFES.DOFH>0
    myLFES = calcJH_KH_AH(myLFES);
end

%% Calculate RefinedTransportation Knowledge Base JH, Constraints Matrix KH, and Transformation Concept AH
if myLFES.DOFH>0
    myLFES = calcJHref_KHref_AHref(myLFES);
end

% Calculate Sequence Independent DOF Measures -- Already Calculated!
myLFES = calcStrucDOF(myLFES);

% Calculate System Knowledge base JS, System Constraints Matrix KS, and System Concept AS
myLFES = calcJS_KS_AS(myLFES);

%% Calculate Heterofunctional Adjacency Matrix AR
myLFES = calcAR(myLFES);


%% Initialize Control Structure within myLFES
myLFES = initializeControl(myLFES);

%% Calculate Controller Agency Matrix
if myLFES.numControllers > 0
    myLFES=makeCAM(myLFES);
end


%% Calculate Controller Adjacency Matrix
if myLFES.numControllers > 0
    myLFES=makeCADM(myLFES);
end

%% Combine DOFS & HFAM with CAM & CADM
if myLFES.numControllers > 0
   myLFES=combineHFAMandCADM(myLFES);
end

%% Calculate Service Graph
if myLFES.numServices > 0
    myLFES=makeServiceGraph(myLFES);
end

%% Calculate Service Feasibility Matrix
if myLFES.numServices > 0
   myLFES=makeServiceFeasibility(myLFES);
end

%% Combine SAM with Services
if myLFES.numServices > 0 && myLFES.numControllers > 0
   myLFES=combineHFAMCADMService(myLFES);
end

%% Mark the LFES Data as Fully Calculated
myLFES.dataState='full';
disp(['I am leaving  ' 'raw2FullLFES.m']);

File: raw2FullLFES/raw2FullLFES.py
from .raw2FullLFES_Functions import *

def raw2FullLFES(LFES,verboseMode):
    '''
    This function processes the output of XML2LFES function and computes HFGT mathematical models.
    :param LFES: Instantiated LFES object with HFGT meta-elements already set-up
    :return: myLFES: Complete LFES, containing HFGT mathematical models
    '''
    print("I am entering raw2FullLFES.py")

    if verboseMode <= 2:
        # Calculate Resource Counts
        calcResourceCounts(LFES)

        # Calculate Resource Indices
        calcResourceIndices(LFES)

        # Calculate Resource Set
        calcResourceSet(LFES)

        # Pack Set of Transformation Processes
        if LFES.DOFM > 0:
            packSetTransformProcess(LFES)

        # Make Set of Transportation Processes
        if LFES.DOFH > 0:
            makeSetTransportProcess(LFES)

        # Make Set of Refined Transportation Processes
        if LFES.DOFH > 0:
            makeSetTransportRefProcess(LFES)
        else:
            LFES.numTransportRefProcess = len(LFES.abstract.methodsxPort.name)*(LFES.numBuffers**2)
            LFES.numHoldingProcess = len(np.unique(LFES.abstract.methodsxPort.ref))

        # Pack Set of Transformation Processes
        if LFES.DOFC > 0:
            packSetControlProcess(LFES)


        # Make Abstract Method Pair Process Index
        if LFES.abstract.methodPair.name1:
            calcMethodPairIdxProc(LFES)

        # Initialize Knowledge Bases and Constraint Matrices
        initializeKnowledgeBasesConstraintsMatrices(LFES)
        initializeKnowledgeBasesConstraintsTensors(LFES)

        # Calculate Transformation Knowledge Base JM, Constraints Matrix KM, and Transformation Concept AM
        if LFES.DOFM > 0:
            calcJM_KM_AM(LFES)

        # Calc Indices
        if LFES.numMachines > 0:
            calcMachineIdxForm(LFES)
            calcMachineIdxPort(LFES)

        if LFES.numIndBuffers > 0:
            calcIndBufferIdxPort(LFES)
        if LFES.numTransporters > 0:
            calcTransporterIdxPort(LFES)

        # Calculate Transportation Knowledge Base JH, Constraints Matrix KH, and Transformation Concept AH
        if LFES.DOFH > 0:
            calcJH_KH_AH(LFES)
            calcJHT_KHT_AHT(LFES)

        # Calculate Refined Transportation Knowledge Base JH, Constraints Matrix KH, and Transformation Concept AH
        if LFES.DOFH > 0:
            calcJHref_KHref_AHref(LFES)
            calcJHrefT_KHrefT_AHrefT(LFES)

        # Calculate Sequence Independent DOF Measures
        calcStrucDOF(LFES)

        # Calculate System Knowledge base JS, System Constraints Matrix KS, and System Concept AS
        calcJS_KS_AS(LFES)

        # Initialize Control Structure within myLFES
        initializeControl(LFES)

        if LFES.numControllers > 0:
            # Calculate Cyber Physical Resource Agency Matrix
            makeCPRAM(LFES)
            # Calculate Cyber Physical Process Agregation Matrix
            makeCPPAM(LFES)
            # Calculate Cyber Physical JS
            calcControllerJS(LFES)
            # Calculate Controller Adjacency Matrix
            makeCADM(LFES)

        # Calculate Service Nets
        if LFES.numServices > 0:
            # makeServiceGraph(LFES)  # was commented out
            makeServiceNet(LFES)

        # Calculate Service Feasibility Matrix
        if LFES.numServices > 0:
            makeServiceFeasibility(LFES)  # Was commented out

        # Calculate Incidence Tensors
        # if verboseMode == 0 or verboseMode == 2:
        calcMLP(LFES)
        calcMRT(LFES)

        # Calculate Hetero-functional Adjacency Matrix AR
        if verboseMode == 0:
            calcARBasic(LFES)
        elif verboseMode == 1:
            calcARPar(LFES)
        elif verboseMode == 2:
            calcARInc(LFES)

        # Combine DOFS & HFAM with CAM & CADM
        if LFES.numControllers > 0:
            if verboseMode == 0:
                combineHFAMandCADM(LFES)

    else:
        # populate only the necesary indices
        methodsToInc(LFES)

        # CalcAR
        print('Entering calcAR')
        calc_t = time.time()
        calcARIncJulia(LFES)
        end_t = time.time()
        print('total time taken calcAR: %f' % (end_t - calc_t))


    # Mark the LFES Data as Fully Calculated
    LFES.dataState = 'full'
    print("I am exiting raw2FullLFES.py")
    return LFES


